### ISSUE TITLE
Indic language maps [name only]

### ISSUE THREAD

**Issue by naveenpf** [osm-in/openstreetmap.in#14]
Need indic langauge maps only for the names. 
http://mlm.jochentopf.com/

No of translations would improve if we can showcase indic language maps.

**Comment by planemad** [comment]
@naveenpf thanks for bringing this up. Any immediate application of this you are seeing?

If we migrate the style from raster to GL, this should be possible. Need to experiment. Reference: https://www.mapbox.com/mapbox-gl-js/example/language-switch/

**Comment by naveenpf** [comment]
@planemad We had offline discussions, what project we have to take up after state highways in Kerala. Few are interested in taking up translation of the map. It would be good to showcase ml map in osm.in, so that we will know the progress and will help to get more volunteers to contribute.My guess is this will help get more volunteers to contribute in all indic languages. 

http://wiki.openstreetmap.org/wiki/Kerala

**Comment by planemad** [comment]
@naveenpf Amazing! Following progress on the [OSM wiki](http://wiki.openstreetmap.org/wiki/Kerala#.E0.B4.AD.E0.B5.82.E0.B4.AA.E0.B4.9F.E0.B4.A8.E0.B4.BF.E0.B5.BC.E0.B4.AE.E0.B5.8D.E0.B4.AE.E0.B4.BE.E0.B4.A3.E0.B4.82). cc @maning

**Comment by planemad** [comment]
Tried my hand at a GL Malayalam map

![untitled2](https://cloud.githubusercontent.com/assets/126868/13582312/b7126e66-e4d1-11e5-87a5-42c9969acb10.gif)

[View](https://api.mapbox.com/styles/v1/planemad/cilb8uyej00fkc4lx4n4fk8uc.html?title=true&access_token=pk.eyJ1IjoicGxhbmVtYWQiLCJhIjoiemdYSVVLRSJ9.g3lbg_eN0kztmsfIPxa9MQ#6.16/10.448/75.406)

- Extracted the data via overpass query `name:ml and place` and created a Studio tileset using [these instructions](https://gist.github.com/planemad/314c953c7827c2466003)
- Created a new Studio style with the tileset, replaced the place label layers to use the uploaded data.

cc @naveenpf

**Comment by naveenpf** [comment]
Thanks a lot @planemad ... excited to see malayalam map :-) ...

**Comment by planemad** [comment]
@naveenpf Can you suggest two Malayalam fonts to use. One with a variety of different weights like bold/light and another visually different to use for seas and oceans.

**Comment by naveenpf** [comment]
I think we can use Rachana Bold and AnjaliOldLipi 
http://smc.org.in/fonts/

**Comment by santhoshtr** [comment]
Note that using good fonts is not enough. As we see in the gif above, the complex script rendering is completely broken. The library that renders the text on the maps should have complex script rendering capability.

**Comment by planemad** [comment]
@naveenpf updated to Rachana.

@santhoshtr yes, complex script rendering is yet to be built in GL maps https://github.com/mapbox/mapbox-gl/issues/4 . Any good example you can point out for a test fixture?

**Comment by santhoshtr** [comment]
> Any good example you can point out for a test fixture?

കോഴിക്കോട് (Kozhikkod/Calicut) should render as 
![image](https://cloud.githubusercontent.com/assets/161672/13593916/d4451354-e525-11e5-81f7-5b5bd320abdf.png)

**Comment by planemad** [comment]
Thanks @santhoshtr . From upstream, seems like we might start to see progress on this by mid April.

**Comment by manuvarkey** [comment]
Complex rendering is fine in mapbox-classic. See this.
[(https://api.mapbox.com/v4/manuvarkey.2267f3e9/page.html?access_token=pk.eyJ1IjoibWFudXZhcmtleSIsImEiOiJjaWdlMTZsd24xZWZ3NnJtNzk2dXhvcjJmIn0.coLgvL8Ct439UzhAmJwI1w#10/11.9836/75.8620)]

@santhoshtr - please see these bug reports for OsmAnd relating to this issue.
[https://github.com/osmandapp/Osmand/issues/2016]
[https://github.com/osmandapp/Osmand/issues/1161]

**Comment by planemad** [comment]
@manuvarkey thats right, the older raster tiles uses [mapnik](http://mapnik.org) rendering which supports complex shaping. Bringing this support to in browser rendering with GL will take some more time.

For now, the best approach might be to allows layer switching between the english tiles and various languages created by individuals. The style+data updates will be the responsibility of the contributor.

**Comment by naveenpf** [comment]
@planemad what is suggestion making this as www.ml.openstreetmap.in with proper India borders ?

**Comment by pnorman** [comment]
> @manuvarkey thats right, the older raster tiles uses mapnik rendering which supports complex shaping. Bringing this support to in browser rendering with GL will take some more time.

I'd suggest avoiding Mapbox GL unless you have internal information that suggests that mapbox/mapbox-gl#4 is close to resolution. Glyph placement, unicode and bidi issues often take a long time to solve, and it's important to have indic scripts rendering right for an Indian site.

With vector tiles like mapbox classic it's still possible to have multiple languages in the vtiles and then just the quick vtile->png step needs to be done for each language.

If you want client-side rendering, I'm not sure if Tangram is any better for bidi support.

**Comment by planemad** [comment]
🚨 Blocked by complex text support in Mapbox GL. Tracking here: https://github.com/mapbox/mapbox-gl-native/issues/7774

**Comment by demonshreder** [comment]
I know GL migration is a goal but is it possible to configure current raster tiles to show indic names through changes in style script?

**Comment by planemad** [comment]
@demonshreder absolutely, given that GL might not get Indic script support soon. Maybe something like https://jsfiddle.net/planemad/Lcvp5p7x/ ?

**Comment by demonshreder** [comment]
This is great. Can this code be copy pasted for anyone to use? Should we be worried about someone using the same API key, if so I can try to replicate this with openlayers.js.

**Comment by planemad** [comment]
Trying to think through the options we have for making a multilingual map.

**Goal** OSM maps with a language selector to change labels to any Indic language

### Client side vector rendering

OSM India has community account vector tile hosting sponsored by Mapbox, so thats a big plus. Its also fairly easy to setup our own [vector tiles server](https://wiki.openstreetmap.org/wiki/Vector_tiles#Server) if needed. This solves the headache of creating the tiles with multilingual data, but rendering options are limited.

- [Mapbox GL](https://docs.mapbox.com/mapbox-gl-js/examples/) would have been preferred due to the extensive capability but Indic script support is still lacking https://github.com/mapbox/mapbox-gl-native/issues/7774  <br>
<img width="113" alt="Screen Shot 2020-06-22 at 2 25 14 PM" src="https://user-images.githubusercontent.com/126868/85322380-412f4d80-b494-11ea-8f49-a4f72bb5cb42.png"><br>Renders as Napolam instead of Nepalam in Tamil
- [Openlayers GL](https://openlayers.org/en/latest/examples/mapbox-layer.html) same issue as Mapbox GL
- [tangram](https://github.com/tangrams/tangram). There are a few demos from the past which had done indic rendering but they all seem down and current development status of the project is unknown  :
  - https://thejeshgn.com/2016/09/24/kannada-web-maps-using-tangram-by-mapzen/ @thejeshgn 
  - https://github.com/tallytalwar/tangrams-indic @tallytalwar @srikanthlogic 


### Server side raster rendering

[Server side tile renderers](https://wiki.openstreetmap.org/wiki/Creating_your_own_tiles) can be cumbersome to setup and are resource intensive, however we can be assured that all the languages will be rendered correctly.

- Mapbox is our current raster tile host on http://openstreetmap.in/ and supports [indic rendering](https://yogiks.github.io/osm-kn/). Unfortunately the raster tile service is [getting deprecated](https://blog.mapbox.com/deprecating-studio-classic-styles-d8892ac38cb4)  ⛔
- [maptiler](https://cloud.maptiler.com). Their raster tile service seems to be using GL based rasterizing which is identical to Mapbox and does not support the indic text correctly.
- Wikimedia tiles: https://jsfiddle.net/planemad/Lcvp5p7x/22/ has map tiles for every language. Downside is that it does not update regularly and we cannot control any of the styling and modify the borders..
- OSM India tile server: Setting our own tile server seems like the only feasible option if we want control over the data updates and any styling ✅ There are a few options for setting this up [openmaptiles](https://github.com/openmaptiles/openmaptiles), [switch2osm](https://switch2osm.org/serving-tiles/using-a-docker-container/)

## Conclusion

It does not seem like there is good solution that will work out of the box for custom styled indic language maps. 

Short term, we could use the Wikimedia indic tiles with a custom boundary overlay https://jsfiddle.net/planemad/Lcvp5p7x/23/ as our multilingual map.

Long term, we should look at setting up our own raster tile server to generate tiles for every indic language. These can probably be just transparent label only tiles which can be overlaid onto any vector basemap. This might be a fairly involved lift and might need some dedicated hands to dig into the technical details. Any volunteers?

**Comment by tallytalwar** [comment]
Thanks for the tag @planemad. tangram (web) is still in active development, tangram-es is more a hobby project from us core developers now.

tangram (webgl) supports all indic fonts, I had extensively tested this with Brett when we were working on it and did a demo for this, if needed I can cook up something using the mapzen stylesheets (PLUS: All this is under Linux Foundation now).

tangram-es supports all indic too with full bidi, complex text - shaping/positioning support, and both tangram web and native renderers will work with the same stylesheet, just in case osm-in plans to have a mobile app! (An example of an app using tangram-es heavily is https://github.com/westnordost/StreetComplete).

Anyways let me know if you have more questions or want any specific help from me.
Good Luck.

**Comment by planemad** [comment]
Thanks @tallytalwar, if tangram is under active development thats a big 👍 to use it. If you can help with getting the [tangram indic demo](https://github.com/tallytalwar/tangrams-indic) running again, that would be a great first step since you would be most knowledgable of what needs to be fixed.

After that it should be fairly trivial to add a language picker to allow users to toggle between options.

**Comment by tallytalwar** [comment]
Great @planemad, thanks for considering.

I will try to spend sometime this week (time is hard to find these days with covid!! :P) and get to this. I am not a web person, so might be pinging the tangram developers if I am stuck with something, but will keep you guys updated.

Also if in the meantime you or anyone wants to play with things on tangram web: https://tangrams.github.io/refill-style/#3.25/20.36/467.42 (https://github.com/tangrams/refill-style).

**Comment by planemad** [comment]
@tallytalwar thanks for the links, forked refill-style to https://github.com/osm-in/map with some modifications and it seems to work!

Demo: https://osm-in.github.io/map/#5.1/23.300/82.000

<img width="881" alt="Screen Shot 2020-06-28 at 2 06 54 AM" src="https://user-images.githubusercontent.com/126868/85939624-1a827400-b8e5-11ea-8947-4ada2d0db608.png">
<img width="866" alt="Screen Shot 2020-06-28 at 2 11 24 AM" src="https://user-images.githubusercontent.com/126868/85939627-1bb3a100-b8e5-11ea-9885-dcb23111fe85.png">
<img width="781" alt="Screen Shot 2020-06-28 at 2 12 55 AM" src="https://user-images.githubusercontent.com/126868/85939628-1ce4ce00-b8e5-11ea-9d0e-1617f6c6f27e.png">

**Issues**
- The [nextzen OSM vector tiles](https://www.nextzen.org) that is used has not been updated since 2017. We would need to replace this with a more updated vector tile source. Changing the source will also require significant map style changes.
  - The tiles do not contain all language labels at all zoom levels.
- Borders need to be fixed as per official ones

**Comment by tallytalwar** [comment]
Great @planemad . Sorry I haven't been able to work on the indic demo rep, but great refill worked great for you.

On nextzen tiles yeah, switching vector tiles would involve a lot of style changes, unless there is an updated service based on https://github.com/tilezen/vector-datasource. Let me ping a few people here and see what's the status on updates.

Do you have any specific question I can relay?

Again, apologies that I havnt been able to get going on the indic demo.

**Comment by tallytalwar** [comment]
And for a playground for tweaking map style you can try tangram-play  (https://github.com/tangrams/tangram-play).

**Comment by tallytalwar** [comment]
Let me send you info on borders too, some work was done to get borders working for different conflicting areas of the world in collaboration with here for nextzen vector tiles.. digging this now.

**Comment by tallytalwar** [comment]
Ok you should be able to follow the instructions from here and make appropriate modifications to the style sheet to reflect the map borders.. https://www.openstreetmap.org/user/nvk/diary/390496

This might be non trivial change so let me know if you need more info here.

**Comment by tallytalwar** [comment]
Also based on the link above for tilezen (source for nextzen vector tiles), its released quarterly, any specific data are you missing, maybe it's not styled?

**Comment by planemad** [comment]
> tangram-play

@tallytalwar Wonderful. Was able to load the style from the url `https://raw.githubusercontent.com/osm-in/map/master/refill-style.yaml` in https://play.tangram.city/

> border fix https://www.openstreetmap.org/user/nvk/diary/390496

Looks good to get started, thanks.
> Also based on the link above for tilezen (source for nextzen vector tiles), its released quarterly

Quarterly is pretty good., https://www.nextzen.org/ seemed to indicated the OSM data was archived from 2017, although i can see changes made after that. There does not seem any specific update frequency mentioned anywhere.

- Tile source: https://github.com/tilezen/vector-datasource
- Tile docs: https://tilezen.readthedocs.io/en/latest/layers/

>  any specific data you are missing, maybe it's not styled?

In lower zoom levels, the place data seems to be coming from natural earth which has limited indic language name coverage (only Hindi). What would be the best way then to make a fully localised map at a country level?
<img width="388" alt="Screen Shot 2020-06-28 at 12 27 48 PM" src="https://user-images.githubusercontent.com/126868/85952953-d0c47880-b93a-11ea-85d5-642710efebbe.png">
<img width="365" alt="Screen Shot 2020-06-28 at 12 30 15 PM" src="https://user-images.githubusercontent.com/126868/85952994-23059980-b93b-11ea-9119-eb0fb06d39bd.png">

**Comment by planemad** [comment]
Success! Got the borders updated https://osm-in.github.io/map/

<img width="653" alt="Screen Shot 2020-06-28 at 1 38 43 PM" src="https://user-images.githubusercontent.com/126868/85955142-b0041f00-b94a-11ea-8164-1856aa697072.png">

**Comment by tallytalwar** [comment]
This is all great, glad all this was helpful to you. Really happy to see all this used in osm-in :). What all is needed next other than wof data and natural earth low zoom data?

**Comment by planemad** [comment]
> What all is needed next other than wof data and natural earth low zoom data?

@tallytalwar Thats seems like the only issue with the tiles. It would be good to know the update frequency and the next release.

On a tangram level, how would one go about implementing dynamic label localisation from Wikidata like https://observablehq.com/@planemad/country-names-in-any-language :
- Get all the tile features with wikidata_id in the map view
- Query Wikidata for localised names for these ids
- Add a dynamic layer to the map and label these features with the localised names
